sklearn/decomposition/_nmf.py

```diff
@@ -705,20 +772,32 @@ def _multiplicative_update_h(X, W, H, beta_loss, l1_reg_H, l2_reg_H, gamma):
         denominator = denominator + l2_reg_H * H
     denominator[denominator == 0] = EPSILON
 
-    numerator /= denominator
-    delta_H = numerator
+    if gamma != 1.:
+        H **= 1. / gamma
+
+    if A is not None and B is not None:
+        A *= rho
+        B *= rho
+        A += numerator * H
+        B += denominator
+        numerator = A
+        denominator = B
+        H = (np.divide(A, B, dtype=X.dtype))
+    else:
+        H *= (np.divide(numerator, denominator, dtype=X.dtype))
 
     # gamma is in ]0, 1]
-    if gamma != 1:
-        delta_H **= gamma
+    if gamma != 1.:
+        H **= gamma
```

Comment:
We should try to avoid unecessary operations. For instance the ``H **= 1. / gamma`` should only be done in the minibatch case. We can also try to do more in-place operations.

```py
if A is not None and B is not None:
    if gamma != 1:
        H **= 1 / gamma
    numerator *= H
    A *= rho
    B *= rho
    A += numerator
    B += denominator
    H = A / B

    if gamma != 1:
        H **= gamma

else:
    delta_H = numerator
    delta_H /= denominator
    if gamma != 1:
        delta_H **= gamma
    H *= delta_H
```